lib/Conversion/TcpToLinalg/DataMovement.cpp

```diff
@@ -0,0 +1,101 @@
+//===------------------------------------------------------------*- C++ -*-===//
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir-tcp/Conversion/TcpToLinalg/TcpToLinalg.h"
+
+#include "mlir-tcp/Dialect/IR/TcpDialect.h"
+#include "mlir-tcp/Dialect/IR/TcpOps.h"
+
+#include "../PassDetail.h"
+#include "PopulatePatterns.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+using namespace mlir::tcp;
+
+namespace {
+
+class ConvertGatherOp : public OpConversionPattern<GatherOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(GatherOp op, OpAdaptor adaptor,
+                                ConversionPatternRewriter &b) const override {
+    Location loc = op->getLoc();
+    auto resultTensorType = getTypeConverter()
+                                ->convertType(op->getResult(0).getType())
+                                .cast<RankedTensorType>();
+
+    auto inputTensor = op->getOperands()[0];
```

Comment:
I've also never quite understood whether its safe to use `op->...` methods or one should use `adaptor.getInput` to account for prior replacements. Might be cleaner to always use `adaptor.getInput` etc. here?

Reply:
My understanding is that` adaptor.getInput` gets the input after all conversions, whereas `op.getInput` gets the original input.

e.g. assuming we have

1% = tcp.sigmoid(xxx)
2% =  tcp.exp(%1)

which lowers to 

1% = linalg.sigmoid(xxx)
2% =  linalg.exp(%1)

For the `exp` op, if we do op.getInput, we will get the value of `tcp.sigmoid`, but adaptor.getInput will get the value of `linalg.sigmod`, which is after other conversions if there is any.

Reply:
That is my understanding too. This makes a big difference especially when the types change during the conversion. In this case, it shouldn't make much of a difference between the two usages.

However, I agree that it is better to use `adaptor` based methods for clarify and consistency. Updated this code.